# 项目：评估和清理英国电商公司销售数据

## 分析目标

此数据分析的目的是，根据市场销售数据，挖掘畅销产品，以便制定更有效的市场策略来提升营收。

本实战项目的目的在于练习评估数据干净和整洁度，并且基于评估结果，对数据进行清洗，从而得到可供下一步分析的数据。

## 简介

原始数据集记录了一家英国在线零售公司在2010年12月1日至2011年12月9日期间的所有交易情况，涵盖了该公司在全球不同国家和地区的业务数据。该公司主要销售覆盖各个场景的礼品，包括但不限于生日礼品、结婚纪念品、圣诞礼品等等。该公司的客户群体主要包括批发商和个人消费者，其中批发商占据了相当大的比例。

数据每列的含义如下：
- `InvoiceNo`: 发票号码。6位数，作为交易的唯一标识符。如果这个代码以字母“c”开头，表示这笔交易被取消。
- `StockCode`: 产品代码。5位数，作为产品的唯一标识符。
- `Description`: 产品名称。
- `Quantity`: 产品在交易中的数量。
- `InvoiceDate`: 发票日期和时间。交易发生的日期和时间。
- `UnitPrice`: 单价。价格单位为英镑（£）。
- `CustomerID`: 客户编号。5位数，作为客户的唯一标识符。
- `Country`: 国家名称。客户所居住的国家的名称。

## 读取数据

In [1]:
import pandas as pd

In [3]:
orignal_data = pd.read_csv("./e_commerce.csv")

In [4]:
orignal_data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice  CustomerID         Country  
0        12/1/2010 8:26       2.55     17850.0  United Kingdom  
1        12/1/2010 8:26       3.39     17850.0  United Kingdom  
2        12/1/2010 8:26       2.75     17850.0  United Kingdom  
3        12/1/2010 8:26       3.39     17850.0  United Kingdom  
4        12/1/2010 8:26       3.39     17850.0  United Kingdom  
...                 ...        ...         ...             ...  
541904  12/9/2011 12:50       0.85     12680.0          France  
541905  12/9/2011 12:50       2.10     12680.0          France  
541906  12/9/2011 12:50       4.15     12680.0          France  
541907  12/9/2011 12:50       4.15     12680.0          France  
541908  12/9/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

## 评估数据

#### 评估数据的结构性混乱

###### 我们可以通过样本提取查看数据是否遵从三原则

In [7]:
orignal_data.sample(10)

InvoiceNo StockCode                        Description  Quantity  \
485119    577554     22796      PHOTO FRAME 3 CLASSIC HANGING         2   
494355    578270     21933    PINK VINTAGE PAISLEY PICNIC BAG         5   
260648    559816     21868               POTTING SHED TEA MUG         1   
425763    573338     22867            HAND WARMER BIRD DESIGN         2   
248436    558835     21890  S/6 WOODEN SKITTLES IN COTTON BAG         1   
10640     537237     22904          CALENDAR PAPER CUT DESIGN         3   
448046    575007     22972            CHILDREN'S SPACEBOY MUG        12   
188515    553041     22521                                NaN        -9   
246441    558700     21935                 SUKI  SHOULDER BAG         1   
275855    561022     17003                BROCADE RING PURSE         36   

             InvoiceDate  UnitPrice  CustomerID         Country  
485119  11/20/2011 15:35       9.95     15533.0  United Kingdom  
494355  11/23/2011 13:39       2.46     14096.0  United Kingdom  
260648   7/12/2011 16:11       3.29         NaN  United Kingdom  
425763  10/30/2011 11:47       2.10     14546.0  United Kingdom  
248436    7/4/2011 11:58       5.79         NaN  United Kingdom  
10640     12/6/2010 9:58       5.91         NaN  United Kingdom  
448046   11/8/2011 11:20       1.65     14140.0  United Kingdom  
188515    5/13/2011 9:01       0.00         NaN  United Kingdom  
246441    7/1/2011 12:33       1.65     17920.0  United Kingdom  
275855   7/24/2011 10:24       0.29     14800.0  United Kingdom

###### 可以看得出来，“每列都对应一个变量，每行一个观察值，每单元格对应一个值”，该数据并不存在结构性问题

#### 评估数据的内容性混乱

##### 评估缺失值

In [8]:
orignal_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


###### 可以看出，在`541909`个数据中，`Description`与`CustomerID`的数据出现了缺失，需要进行提取具体查看是否会对畅销产品的查找造成影响
###### 而且`InvoiceDate`应为DataTime，`CustomerID`应为字符串

In [16]:
orignal_data[orignal_data["Description"].isnull()]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
622       536414     22139         NaN        56  12/1/2010 11:52        0.0   
1970      536545     21134         NaN         1  12/1/2010 14:32        0.0   
1971      536546     22145         NaN         1  12/1/2010 14:33        0.0   
1972      536547     37509         NaN         1  12/1/2010 14:33        0.0   
1987      536549    85226A         NaN         1  12/1/2010 14:34        0.0   
...          ...       ...         ...       ...              ...        ...   
535322    581199     84581         NaN        -2  12/7/2011 18:26        0.0   
535326    581203     23406         NaN        15  12/7/2011 18:31        0.0   
535332    581209     21620         NaN         6  12/7/2011 18:35        0.0   
536981    581234     72817         NaN        27  12/8/2011 10:33        0.0   
538554    581408     85175         NaN        20  12/8/2011 14:06        0.0   

        CustomerID         Country  
622            NaN  United Kingdom  
1970           NaN  United Kingdom  
1971           NaN  United Kingdom  
1972           NaN  United Kingdom  
1987           NaN  United Kingdom  
...            ...             ...  
535322         NaN  United Kingdom  
535326         NaN  United Kingdom  
535332         NaN  United Kingdom  
536981         NaN  United Kingdom  
538554         NaN  United Kingdom  

[1454 rows x 8 columns]

###### 这些抽样数据中呈现出的特征都是`UnitPrice`为零，即便具不具备`Description`都无法呈现出畅销产品应具备的特征。
###### 但是否所有缺失`Description`的数据都是`UnitPrice`为零呢？这需要进一步的提取。

In [17]:
orignal_data[(orignal_data["Description"].isnull()) & (orignal_data["UnitPrice"] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

###### 上述的零数据表明，所有缺失`Description`的数据都是`UnitPrice`为零。因此缺失`Description`数据可以直接清洗

###### 接下来查看`CustomerID`

###### 虽然`CustomerID`存在缺失，但`CustomerID`与`Country`指代的是同一事物，且`Country`比`CustomerID`更加直观，因此Series`CustomerID`可以直接删除

##### 评估重复值

In [25]:
orignal_data[orignal_data.duplicated()]

InvoiceNo StockCode                        Description  Quantity  \
517       536409     21866        UNION JACK FLAG LUGGAGE TAG         1   
527       536409     22866      HAND WARMER SCOTTY DOG DESIGN         1   
537       536409     22900    SET 2 TEA TOWELS I LOVE LONDON          1   
539       536409     22111       SCOTTIE DOG HOT WATER BOTTLE         1   
555       536412     22327  ROUND SNACK BOXES SET OF 4 SKULLS         1   
...          ...       ...                                ...       ...   
541675    581538     22068        BLACK PIRATE TREASURE CHEST         1   
541689    581538     23318     BOX OF 6 MINI VINTAGE CRACKERS         1   
541692    581538     22992             REVOLVER WOODEN RULER          1   
541699    581538     22694                       WICKER STAR          1   
541701    581538     23343       JUMBO BAG VINTAGE CHRISTMAS          1   

            InvoiceDate  UnitPrice  CustomerID         Country  
517     12/1/2010 11:45       1.25     17908.0  United Kingdom  
527     12/1/2010 11:45       2.10     17908.0  United Kingdom  
537     12/1/2010 11:45       2.95     17908.0  United Kingdom  
539     12/1/2010 11:45       4.95     17908.0  United Kingdom  
555     12/1/2010 11:49       2.95     17920.0  United Kingdom  
...                 ...        ...         ...             ...  
541675  12/9/2011 11:34       0.39     14446.0  United Kingdom  
541689  12/9/2011 11:34       2.49     14446.0  United Kingdom  
541692  12/9/2011 11:34       1.95     14446.0  United Kingdom  
541699  12/9/2011 11:34       2.10     14446.0  United Kingdom  
541701  12/9/2011 11:34       2.08     14446.0  United Kingdom  

[5268 rows x 8 columns]

###### 重复值多出现在`CustomerID`与`Country`中，这种重复值并不会影响畅销产品的查找，因此可以不予处理
###### 即便有`InvoiceNo`、`StockCode`、`Description`的数据出现重复，但其可以理解为多个订单，因此可以不予处理

##### 评估不一致值

###### 此评估旨在找出不同值但表示同一现实对象的值
###### 在每个列名中，唯独`Country`很有可能出现不同值但表示同一现实对象的值，如"UK""United Kingdom"都表示英国，因此要予以评估

###### 可以先通过对每个出现的值的个数进行总结，在查找是否出现不同值但表示同一现实对象的值

In [32]:
orignal_data["Country"].value_counts()

Country
United Kingdom          495266
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
China                      288
Singapore                  229
USA                        218
UK                         211
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United States               73
United Arab Emirates        68


###### 这里可以发现`United Kingdom`、`UK`、`U.K.`都是指代英国，`USA`、`United States`都是指代美国，这无疑需要在清洗数据中进行统一

##### 评估无效或错误值

In [33]:
orignal_data.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

###### 令人惊讶的是，`Quantity`数量与`UnitPrice`单价竟然出现了负数，我们应该进行提取然后分析是否会对畅销产品的查找造成问题

###### 首先查看`Quantity`为负数的数据

In [34]:
orignal_data[orignal_data["Quantity"] < 0]

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

            InvoiceDate  UnitPrice  CustomerID         Country  
141      12/1/2010 9:41      27.50     14527.0  United Kingdom  
154      12/1/2010 9:49       4.65     15311.0  United Kingdom  
235     12/1/2010 10:24       1.65     17548.0  United Kingdom  
236     12/1/2010 10:24       0.29     17548.0  United Kingdom  
237     12/1/2010 10:24       0.29     17548.0  United Kingdom  
...                 ...        ...         ...             ...  
540449   12/9/2011 9:57       0.83     14397.0  United Kingdom  
541541  12/9/2011 10:28     224.69     15498.0  United Kingdom  
541715  12/9/2011 11:57      10.95     15311.0  United Kingdom  
541716  12/9/2011 11:58       1.25     17315.0  United Kingdom  
541717  12/9/2011 11:58       1.25     17315.0  United Kingdom  

[10624 rows x 8 columns]

###### 可以看到，抽样出来`Quantity`为负数的数据，都具有`InvoiceNo`含"C"的情况（即被取消订单），那么是否所有`Quantity`为负数的数据都是被取消订单呢？

In [36]:
orignal_data[(orignal_data["Quantity"] < 0) & (orignal_data["InvoiceNo"].str[0] !="C")]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
2406      536589     21777         NaN       -10  12/1/2010 16:50        0.0   
4347      536764    84952C         NaN       -38  12/2/2010 14:42        0.0   
7188      536996     22712         NaN       -20  12/3/2010 15:30        0.0   
7189      536997     22028         NaN       -20  12/3/2010 15:30        0.0   
7190      536998     85067         NaN        -6  12/3/2010 15:30        0.0   
...          ...       ...         ...       ...              ...        ...   
535333    581210     23395       check       -26  12/7/2011 18:36        0.0   
535335    581212     22578        lost     -1050  12/7/2011 18:38        0.0   
535336    581213     22576       check       -30  12/7/2011 18:38        0.0   
536908    581226     23090     missing      -338   12/8/2011 9:56        0.0   
538919    581422     23169     smashed      -235  12/8/2011 15:24        0.0   

        CustomerID         Country  
2406           NaN  United Kingdom  
4347           NaN  United Kingdom  
7188           NaN  United Kingdom  
7189           NaN  United Kingdom  
7190           NaN  United Kingdom  
...            ...             ...  
535333         NaN  United Kingdom  
535335         NaN  United Kingdom  
535336         NaN  United Kingdom  
536908         NaN  United Kingdom  
538919         NaN  United Kingdom  

[1336 rows x 8 columns]

###### 事实并非如此，所有`Quantity为`负数的数据只有部分是被取消订单，但同时又可以发现，这些不畏被取消订单的数据：`UnitPrice`为零，这又不会影响畅销产品的查找。
###### 所以我们可以增加条件，看是否可以排除`Quantity`为负数的所有数据

In [39]:
orignal_data[(orignal_data["Quantity"] < 0) & (orignal_data["InvoiceNo"].str[0] !="C") & (orignal_data["UnitPrice"] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

###### 这下没有了！也就是说，`Quantity`为负数的数据，不是被取消订单，就是单价为零，无法造成对畅销产品的影响，因此可以被清洗掉。

###### 我们再继续查看`UnitPrice`为负数的数据

In [40]:
orignal_data[orignal_data["UnitPrice"] < 0]

InvoiceNo StockCode      Description  Quantity      InvoiceDate  \
299983   A563186         B  Adjust bad debt         1  8/12/2011 14:51   
299984   A563187         B  Adjust bad debt         1  8/12/2011 14:52   

        UnitPrice  CustomerID         Country  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom

###### 此为坏账数据，不会对畅销产品的查找造成影响，可以大胆清洗

## 清理数据

#### 根据数据的评估，姑且可以做出以下清洗步骤：
###### 将`InvoiceDate`变为datatime，`CustomerID`变为字符串
###### 直接清洗缺失`Description`的数据
###### 将不一致值进行统一
###### 清洗`Quantity`、`UnitPrice`为负数的数据

In [65]:
cleaned_data = orignal_data.copy()

##### 将`InvoiceDate`变为datatime，`CustomerID`变为字符串

In [47]:
cleaned_data["InvoiceDate"] = pd.to_datetime(cleaned_data["InvoiceDate"])
cleaned_data["InvoiceDate"]

0        2010-12-01 08:26:00
1        2010-12-01 08:26:00
2        2010-12-01 08:26:00
3        2010-12-01 08:26:00
4        2010-12-01 08:26:00
                 ...        
541904   2011-12-09 12:50:00
541905   2011-12-09 12:50:00
541906   2011-12-09 12:50:00
541907   2011-12-09 12:50:00
541908   2011-12-09 12:50:00
Name: InvoiceDate, Length: 541909, dtype: datetime64[ns]

In [49]:
cleaned_data["CustomerID"].astype(str)

0         17850.0
1         17850.0
2         17850.0
3         17850.0
4         17850.0
           ...   
541904    12680.0
541905    12680.0
541906    12680.0
541907    12680.0
541908    12680.0
Name: CustomerID, Length: 541909, dtype: object

In [66]:
cleaned_data["CustomerID"] = cleaned_data["CustomerID"].slice(0,-2)

AttributeError: 'Series' object has no attribute 'slice'

##### 直接清洗缺失`Description`的数据

In [70]:
cleaned_data.dropna(subset=["Description"],inplace=True)
cleaned_data["Description"].isnull().sum()

np.int64(0)

##### 将不一致值进行统一

In [72]:
cleaned_data["Country"] = cleaned_data["Country"].replace(["UK","U.K."],"United Kingdom")
cleaned_data["Country"] = cleaned_data["Country"].replace("USA","United States")

In [73]:
cleaned_data["Country"].value_counts()

Country
United Kingdom          494024
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
United States              291
China                      288
Singapore                  229
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United Arab Emirates        68
European Community          61
RSA                         58


##### 清洗`Quantity`、`UnitPrice`为负数的数据

In [83]:
cleaned_data = cleaned_data[cleaned_data["Quantity"] >= 0]
cleaned_data[cleaned_data["Quantity"] < 0]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

In [85]:
cleaned_data = cleaned_data[cleaned_data["UnitPrice"] >= 0]
cleaned_data[cleaned_data["UnitPrice"] < 0]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

## 保存清理后的数据

In [87]:
cleaned_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 530691 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    530691 non-null  object 
 1   StockCode    530691 non-null  object 
 2   Description  530691 non-null  object 
 3   Quantity     530691 non-null  int64  
 4   InvoiceDate  530691 non-null  object 
 5   UnitPrice    530691 non-null  float64
 6   CustomerID   397924 non-null  float64
 7   Country      530691 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 36.4+ MB


In [89]:
cleaned_data.describe()

Quantity      UnitPrice     CustomerID
count  530691.000000  530691.000000  397924.000000
mean       10.605855       3.903303   15294.315171
std       156.638147      35.896047    1713.169877
min         1.000000       0.000000   12346.000000
25%         1.000000       1.250000   13969.000000
50%         3.000000       2.080000   15159.000000
75%        10.000000       4.130000   16795.000000
max     80995.000000   13541.330000   18287.000000

In [91]:
cleaned_data.to_csv("e_commerce_cleaned.csv",index=False)

In [92]:
pd.read_csv("e_commerce_cleaned.csv")

C:\Users\Suliko\AppData\Local\Temp\ipykernel_25432\155800971.py:1: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  pd.read_csv("e_commerce_cleaned.csv")


InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
530686    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
530687    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
530688    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
530689    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
530690    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice  CustomerID         Country  
0        12/1/2010 8:26       2.55     17850.0  United Kingdom  
1        12/1/2010 8:26       3.39     17850.0  United Kingdom  
2        12/1/2010 8:26       2.75     17850.0  United Kingdom  
3        12/1/2010 8:26       3.39     17850.0  United Kingdom  
4        12/1/2010 8:26       3.39     17850.0  United Kingdom  
...                 ...        ...         ...             ...  
530686  12/9/2011 12:50       0.85     12680.0          France  
530687  12/9/2011 12:50       2.10     12680.0          France  
530688  12/9/2011 12:50       4.15     12680.0          France  
530689  12/9/2011 12:50       4.15     12680.0          France  
530690  12/9/2011 12:50       4.95     12680.0          France  

[530691 rows x 8 columns]